### ISSUE TITLE
Question regarding manual configuration of dataSources and read-replica selection

### ISSUE THREAD

**Issue by RovoMe** [spring-cloud/spring-cloud-aws#194]
According to the [documentation](http://cloud.spring.io/spring-cloud-aws/spring-cloud-aws.html#_configuring_data_source_with_java_config) Spring cloud AWS uses `@EnableRdsInstance(..., readReplicaSupport = true)` annotation to configure Spring to use a read-replica MySQL instance for `SELECT` queries if `@Transactional(readOnly = true)` is defined.

As `@EnableRdsInstance` annotation requires the some AWS RDS credentials from being defined as annotation properties (i.e.`dbInstanceIdentifier` and `password`) and we do not want to expose these at code level, is there an option to configure both (or multiple) datasources directly with a Java-configuration approach and still make use of the auto-configuring done by spring-cloud-aws to enable the read-replica selection for `@Transactional(readOnly = true)` annotated methods?

A code sample could look like this:

```
@Configuration
@EnableRdsInstance(readReplicaSupport = true) // no dbInstanceIdentifier and password credentials here!
public class MySqlSpringReadReplicaConfig {

    @Bean
    public DataSource springReadReplicaSupportedDataSource() throws Exception {
        ReadOnlyRoutingDataSource roDataSource = new ReadOnlyDataSource();

        Map<Object, Object> targetDataSource = new HashMap<>();
        targetDataSources.put("read", readReplicaDataSource());

        roDataSource.setTargetDataSources(targetDataSource());
        roDataSource.setDefaultTargetDataSource(writeDataSource());
        roDataSource.afterPropertiesSet();

        return new LazyConnectionDataSourceProxy(roDataSource);
    }

    private DataSource writeDataSource() {
        HikariConfig config = new HikariConfig();
        ...
        // jdbc URL will be injected from encrypted properties directly
        ...
        return new HikariDataSource(config);
    }

    private DataSource readReplicaDataSource() {
        HikariConfig config = new HikariConfig();
        ...
        // jdbc URL will be injected from encrypted properties directly
        ...
        return new HikariDataSource(config);
    }
} 
```

The MySQL instances are hosted on AWS RDS though we configure our DataSources on our own via a `jdbcUrl` field on the data source directly instead of wanting to rely on some hidden magic to build this URL for us. Though we want to enable the automatic lookup of the read-replica for `@Transactionl(readOnly = true)` annotated methods.

If this is not directly possible with spring-cloud-aws, is there an other spring solution which offers this kind of feature?

**Comment by aemruli** [comment]
@RovoMe there is no need to have password and identifier int he code. You can use placeholder or expression to access the values. Nevertheless, you can in theory use the code for read-replicas without using the data source that is built by spring cloud aws. 
In this case you will need to configure your data source along with org.springframework.cloud.aws.jdbc.datasource.ReadOnlyRoutingDataSource which could be easily done in a java config class. org.springframework.cloud.aws.jdbc.datasource.ReadOnlyRoutingDataSource#setTargetDataSources takes the read-replicas and org.springframework.cloud.aws.jdbc.datasource.ReadOnlyRoutingDataSource#setDefaultTargetDataSource takes your data source which is the master.

**Comment by RovoMe** [comment]
I managed ot get read-only annotated service issuing SELECT queries against the replica database to work by making use of `ReadOnlyRoutingDatasource`

**Comment by bjoernhaeuser** [comment]
@RovoMe do you mind sharing your code? :9

**Comment by RovoMe** [comment]
@bjoernhaeuser Well, we disbandoned SCA as [we only used one single class](https://github.com/spring-cloud/spring-cloud-aws/blob/master/spring-cloud-aws-jdbc/src/main/java/org/springframework/cloud/aws/jdbc/datasource/ReadOnlyRoutingDataSource.java) of it, which we refactored into our own version by adding some log-statements (nothing more) to see when the read-replica kicks in and when the write datasource is used.

The whole code is not publicly available though I can try to copy & paste the main parts in here. We basically have 2 Spring configurations, one for the read+write data source and one for the read-replica data source. While the configuration for the read+write data source contains most of the bean declarations, the read-replica Spring config "overrides" these bean definitions and replaces the read+write datasource with the read-replica datasource. This is needed in our case as our external testing platform does not yet support read-replication as it runs on a minimal setup (only one MySQL database, no replication):

```java
/**
 * This Spring configuration defines a datasource that is capable of directing 
 * <em>SELECT</em> SQL queries that are marked as read-only to a read replicated 
 * MySQL database instead of using the datasource defined for read/write access.
 * <p/>
 * In order to achieve this capabilities this configuration does import some bean 
 * definitions defined in {@link DatabaseWriteSpringConfig} and overrides them if they 
 * directly access the read/write datasource.
 * <p/>
 * To make use of reads issued against the read-replica datasource annotate methods 
 * issuing SQL SELECT queries with <em>@Transactional(readOnly = true)</em> and 
 * import this configuration instead of {@link DatabaseWriteSpringConfig}. 
 * <em>SELECT</em> queries not marked as <em>readOnly = true</em> or any other 
 * SQL requests will be issued against the read/write datasource.
 */
@Configuration
@Import(DatabaseWriteSpringConfig.class)
@Profile({"prod", "dev", "docker"})
public class DatabaseReadReplicaSpringConfig {

  private static final Logger LOG = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());

  @Resource(name = "messageDataSource")
  private DataSource dataSource;

  @Resource(name = "hibernateProperties")
  private Properties hibernateProperties;

  @Resource
  private Environment env;

  private DataSource readReplicaDataSource() throws Exception {
    // Custom Hikari JDBC Pool to access MySQL replication database whose slaves are intended for
    // read-only access
    HikariConfig config = new HikariConfig();

    try {
      config.setPoolName(
          "HikariPool-HUB-ReadReplica-Pool-" + RandomStringUtils.randomAlphanumeric(5));
    } catch (Exception ex) {
      LOG.warn("Could not set Hikari read-replica pool name: {}", ex.getMessage());
    }

    // Note: we do not want master/slave replication with Connector/J as AWS MySQL replication also
    // does not support this
    config.setJdbcUrl(env.getProperty("jdbc.replicationUrl"));

    config.setUsername(env.getProperty("jdbc.userRead"));
    config.setPassword(env.getProperty("jdbc.passwordRead"));

    config.setDriverClassName("com.mysql.jdbc.Driver");

    // Set socket/connect timeout on jdbc level
    Properties dsProperties = new Properties();
    dsProperties.put("socketTimeout", env.getProperty("jdbc.socketTimeout", Integer.class));
    dsProperties.put("connectTimeout", env.getProperty("jdbc.connectTimeout", Integer.class));
    config.setDataSourceProperties(dsProperties);

    config.setReadOnly(true);
    config.setAutoCommit(false);
    config.setValidationTimeout(5000);
    config.setLeakDetectionThreshold(10000); // 10 seconds

    config.setMinimumIdle(2);
    config.setMaximumPoolSize(20);

    config.setIdleTimeout(600000); // 10 minutes
    config.setMaxLifetime(1800000); // 30 minutes
    config.setConnectionTimeout(30000); // 30 seconds

    config.setInitializationFailFast(false);
    config.setAllowPoolSuspension(false);
    config.setRegisterMbeans(true);

    return new HikariDataSource(config);
  }

  @Bean(name = "readReplicaDataSource")
  public DataSource springReadReplicaDataSource() throws Exception {
    LOG.info("Initializing read-replica enabled datasource");
    ReadOnlyRoutingDataSource roDataSource = new ReadOnlyRoutingDataSource();
    // define the target data sources to read from
    Map<Object, Object> targetDataSources = new HashMap<>();
    targetDataSources.put("read1", readReplicaDataSource());
    roDataSource.setTargetDataSources(targetDataSources);
    // if no read replica is available the default datasource will be taken
    roDataSource.setDefaultTargetDataSource(dataSource);
    roDataSource.afterPropertiesSet();
    return new LazyConnectionDataSourceProxy(roDataSource);
  }

  @Bean(name = "dbTransactionManager")
  public PlatformTransactionManager transactionManager() throws Exception {
    JpaTransactionManager transactionManager = new JpaTransactionManager(entityManagerFactory());
    transactionManager.setDataSource(springReadReplicaDataSource());
    transactionManager.setJpaDialect(new HibernateJpaDialect());
    return transactionManager;
  }

  @Bean(name = "namedJdbcTemplate")
  public NamedParameterJdbcTemplate jdbcTemplate() throws Exception {
    return new NamedParameterJdbcTemplate(springReadReplicaDataSource());
  }

  @Bean(name = "entityManagerFactory")
  public EntityManagerFactory entityManagerFactory() throws Exception {
    LocalContainerEntityManagerFactoryBean
        entityManagerFactory =
        new LocalContainerEntityManagerFactoryBean();
    entityManagerFactory.setDataSource(springReadReplicaDataSource());
    entityManagerFactory.setPackagesToScan("com.your.package.to.scan");
    entityManagerFactory.setPersistenceProvider(new HibernatePersistenceProvider());
    entityManagerFactory.setPersistenceUnitName("DB");
    entityManagerFactory.setPersistenceXmlLocation("classpath:META-INF/persistence-config.xml");
    if (hibernateProperties == null) {
      LOG.warn("Hibernate Properties are null!");
    } else {
      entityManagerFactory.setJpaProperties(hibernateProperties);
    }
    entityManagerFactory.afterPropertiesSet();
    return entityManagerFactory.getObject();
  }
}
```

The actual application Spring config now imports both database Spring configurations, with the read-write configuration first and the read-replica last as we only want the read-replication setup for selected profiles:

```java
@Configuration
@Import({ ...
          DatabaseWriteSpringConfig.class,
          DatabaseReadReplicaSpringConfig.class,
          ...
})
@EnableConfigurationProperties({SomeSettings.class, ...})
@PropertySource("classpath:application-${spring.profiles.active}.properties")
public class ApplicationSpringConfig {
   ...
}
```

If you don't have to distinguish between different setups as all profiles support read-replication you can remove the `@Profile(...)` annotation and also remove the import of the `DatabaseWriteSpringConfig` class as this is imported by the read replica config automatically.

In order to make use of the read-replication a service needs to declare itself (or its methods) as read only. This can be achived via this (incomplete) code sample.

```java
@Service
@Transactional(transactionManager = "dbTransactionManager", readOnly = true)
public class IndexingService {

  private class EntityProjectionResultTransformer implements ResultTransformer {

    @Override
    public Object transformTuple(Object[] objects, String[] strings) {
      SomeEntity se = new SomeEntity();
      se.setId((Long) objects[0]);
      ...

      OtherEntity oe = new OtherEntity();
      oe.setId((Long) objects[n]);
      oe.setOtherValue((String) objects[n+1]);
    }
  }

  @Resource
  private SomeEntityRepository repository;
  @PersistenceContext(unitName = "DB")
  private EntityManager entityManager;

  private String profile;

  @PostConstruct
  public void initialize() {
    ProfileUtils profileUtils = new ProfileUtils(environment.getActiveProfiles());
    profile = profileUtils.getActiveProfile();
  }

  /**
   * Prepares entities to get indexed. If a <em>days</em> parameter is provided will lookup
   * all entities for the past couple of days. If <em>null</em> is provided will index only the
   * last entity.
   * <p/>
   * Note: this method makes use of a scrolling result set in order to reduce memory footprint
   *       while loading data from the DB.
   *
   * @param days The number of past days to include entities for indexing. If <em>null</em> is 
   *             passed only the most recent entity is indexed
   *
   * @see http://knes1.github.io/blog/2015/2015-10-19-streaming-mysql-results-using-java8-streams-and-spring-data.html
   */
  public void indexEntites(Integer days) {
    Date fromDate = null;
    if (days == null) {
      OtherEntity last = repository.findLastEntity();
      if (last != null) {
        Calendar cal = Calendar.getInstance();
        cal.setTime(last.getLastStatusChange());
        fromDate = cal.getTime();
      }
    } else {
      Calendar cal = Calendar.getInstance();
      cal.add(Calendar.DAY_OF_YEAR, days * -1);
      fromDate = cal.getTime();
    }

    if (fromDate == null) {
      LOG.info("Index entity_index is empty - supply a startdate for re-indexing");
      return;
    }

    ScrollableResults scroll = null;

    try {
      Long entityToIndexCount = getEntitiesToIndexCount(fromDate);
      LOG.info("Indexing {} entities", entitiesToIndexCount);
      logProgress(0f);

      scroll = createScrollCriteria(fromDate);

      Map<String, OtherEntity> entitiesMap = new HashMap<>();
      EntityConverter entityConverter = new EntityConverter();
      ScrollableResultIterator<SomeEntity> iterator =
          new ScrollableResultIterator<>(scroll, SomeEntity.class);
      Date lastStatusChangeFrame = null;

      while (iterator.hasNext()) {
        SomeEntity se = iterator.next();
        OtherEntity oe = entitiesMap.get(se.getOtherEntity().getId());
        ...
      }
    } catch (Exception e) {
        LOG.error(e.getMessage(), e);
    } finally {
        if (scroll !=  null) {
           scroll.close();
        }
    }
  }

  public ScrollableResults createScrollCriteria(Date fromDate) {
    Session session = entityManager.unwrap(Session.class);
    int fetchSize = Integer.MIN_VALUE;

    // for some unknown reason, the mysql java driver needs Integer.MIN_VALUE to activate
    // streaming support. h2 throws an exception for this value...
    // http://www.numerati.com/2012/06/26/reading-large-result-sets-with-hibernate-and-mysql/
    if (profile != null && "ci".equals(profile) || "local".equals(profile)) {
      fetchSize = 500;
    }

    ProjectionList projections = Projections.projectionList();
    ...

    Criteria criteria = session.createCriteria(SomeEntity.class, "entity");

    criteria.setReadOnly(true);
    criteria.setCacheable(false);
    criteria.setFetchSize(fetchSize);
    criteria.add(Restrictions.ge("entity.lastStatusChange", fromDate));
    criteria.createAlias("entity.otherEntity", "otherEntity", JoinType.LEFT_OUTER_JOIN);
    criteria.addOrder(Order.asc("entity.lastStatusChange"));
    criteria.setProjection(projections);
    criteria.setResultTransformer(new EntityProjectionResultTransformer());

    return criteria.scroll(ScrollMode.FORWARD_ONLY);
  }
}
```
As soon as Spring sees a read-only method that is accessing either the entity manager or a repository it will use one of the target datasources defined in the `ReadOnlyRoutingDataSource` configured beforehand. If either no read-only access is found or no read-replica datasource is available the configured data soruce should fallback onto the write-enabled data source.

Final Note: if you have the New Relic agent in your classpath this solution will not work as the `TransactionSynchronizationManager.isCurrentTransactionReadOnly()` method will return the wrong result for whatever reason! In that case the read-only datasource will be ignored and the write-enabled datasource will be used due to the fallback mechanic.

**Comment by anuraggupta** [comment]
@bjoernhaeuser @RovoMe Thanks for the post. Now I finally know why the RoutingDataSource was  working in dev setup but not when I deployed on an environment(We use new relic in deployment).

Did New Relic reply ? or Did you guys found a solution for it to work with new relic.

**Comment by bjoernhaeuser** [comment]
Never reported it to them :(

**Comment by anuraggupta** [comment]
I reported it to New Relic and they say they have fixed in version 4.0.0 . I have tried in 4.0.1 and it works now.
